pkg/tools/opautils/opainput.go

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2020 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Now we can consider the OPA input as a map with the keys:
+//    1) "connection" -- we can use all information about connection in OPA (e.g. input.connection.path.path_segments[0].token)
+//    2) "auth_info" with sub keys:
+//          i) "certificate" -- is a pem encoded x509cert (usage: input.auth_info.certificate)
+//			ii) "spiffe_id" -- is a spiffeID from SVIDX509Certificate (usage: input.auth_info.spiffe_id)
+//    3) "operation" -- one of request/close (usage: input.operation)
+//    4) "role" -- one of client/endpoint (usage: input.role)
+
+// An example of using OPA input for the case of token signature verification:
+//
+//		package test
+//
+//		default allow = false
+//
+//		allow {
+//			token := input.connection.path.path_segments[0].token
+//			cert := input.auth_info.certificate
+//			io.jwt.verify_es256(token, cert) # signature verification
+//		}
+//
+
+// Package opautils provides of utilities for using OPA
+package opautils
+
+import (
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/spiffeutils"
+
+	"github.com/pkg/errors"
+
+	"google.golang.org/grpc/credentials"
+)
+
+type role string
+type operation string
+
+const (
+	// Client is a client role
+	Client role = "client"
+
+	// Endpoint is an endpoint role
+	Endpoint role = "endpoint"
+
+	// Request is a request operation
+	Request operation = "request"
+
+	// Close is a close operation
+	Close operation = "close"
+)
+
+// PreparedOpaInput - returns a prepared input for using in OPA
+func PreparedOpaInput(connection *networkservice.Connection, authInfo credentials.AuthInfo, operation operation, role role) (map[string]interface{}, error) {
+	connectionAsMap, err := convertConnectionToMap(connection)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Cannot convert connection %v to map", connection)
+	}
+
+	cert := parseX509Cert(authInfo)
+
+	spiffeID, err := spiffeutils.SpiffeIDFromX509(cert)
+	if err != nil {
+		return nil, err
+	}
+
+	pemcert := pemEncodingX509Cert(cert)
+
+	rv := map[string]interface{}{
+		"connection": connectionAsMap,
+		"auth_info": map[string]interface{}{
+			"certificate": pemcert,
+			"spiffe_id":   spiffeID,
+		},
+		"operation": operation,
+		"role":      role,
```

Comment:
Tell me what you are thinking about the "role" parameter :)

Reply:
So I don't understand what the authorization client chain element should do. Please can you explain some usage cases when we need to use authorization client? Authz server is handle policies by using connection and tls info, but what authz client should do?

Reply:
Client-side authorization is used to decide if you trust who is providing you the Network Service.

Client-side authorization would also use OPA to do authorization based on the TLSInfo and Connection... but in this case its the Connection *returned* from its Request call.